helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixManager.java

```diff
@@ -699,19 +699,15 @@ void createClient() throws Exception {
     int retryCount = 0;
     while (retryCount < 3) {
       try {
-        // TODO: get session id from waitUntilConnected to avoid race condition
-        if (!_zkclient.waitUntilConnected(_connectionInitTimeout, TimeUnit.MILLISECONDS)) {
-          throw new ZkTimeoutException(
-              "Unable to connect to zookeeper server within timeout: " + _connectionInitTimeout
-                  + " ms.");
-        }
+        final long sessionId =
```

Comment:
line 698 _zkclient.subscribeStateChanges(this); 
at this implementation should always happen later than first synconnected for the reason that ZkClient constructor would call waitTillConnected to guarantee the first synconnect is delivered before proceeding. 

Thus first handleNewSession is not delivered from first Synconnect.

By the time handlNewSession is called synchronously from here, it does not necessarily have a valid sessionID. Here, we use waitForEstablishedSession to get a valid sessionID at the time. Otherwise, you have chance to get sessionID 0.  

@jiajunwang, can you confirm if my reasoning is right here?

Reply:
The other issue identified with HuiZhi is that 698 _zkclient.subscribeStateChanges(this) is not safe for another reason. 

The event can be like this:
1/ Zookeeper (client) created, and syconnect comes and then the constructor of ZKClient finishes.
2/ line 698 _zkclient.subscribeStateChanges(this) executed.
3/ Zookeeper expired and reconnect happens in ZKClient, then synconnnected get delivered.
4/ HandleNewSesssion get delivered.
5/ _zkCllient.waitForEstablishedSession executed and handleNewSession get called the second time. 

So the gist is still we should provide a constructor of ZKCLient with the stateChanges callback register at one shot. @jiajunwang what is your take?

Reply:
As we discussed offline, the issue is not newly introduced in this PR. And it is not becoming worse. I agree that we need to fix it. But not in this PR.